lib/ansible/plugins/loader.py

```diff
@@ -388,24 +484,28 @@ def _find_plugin(self, name, mod_type='', ignore_deprecated=False, check_aliases
 
         return None
 
-    def find_plugin(self, name, mod_type='', ignore_deprecated=False, check_aliases=False):
+    def find_plugin(self, name, mod_type='', ignore_deprecated=False, check_aliases=False, collection_list=None):
         ''' Find a plugin named name '''
 
         # Import here to avoid circular import
         from ansible.vars.reserved import is_reserved_name
 
-        plugin = self._find_plugin(name, mod_type=mod_type, ignore_deprecated=ignore_deprecated, check_aliases=check_aliases)
+        plugin = self._find_plugin(name, mod_type=mod_type, ignore_deprecated=ignore_deprecated, check_aliases=check_aliases, collection_list=collection_list)
         if plugin and self.package == 'ansible.modules' and name not in ('gather_facts',) and is_reserved_name(name):
             raise AnsibleError(
                 'Module "%s" shadows the name of a reserved keyword. Please rename or remove this module. Found at %s' % (name, plugin)
             )
 
         return plugin
 
-    def has_plugin(self, name):
+    def has_plugin(self, name, collection_list=None):
         ''' Checks if a plugin named name exists '''
 
-        return self.find_plugin(name) is not None
+        try:
+            return self.find_plugin(name, collection_list=collection_list) is not None
+        # TODO: limit this to only plugin load/resolution errors
+        except Exception:
```

Comment:
At least for alpha/beta releases and for cases where collections are involved, this bare exception should probably at least warn about failures. Otherwise this could silently ignore a lot of problems.


Maybe something like this (untested...)

``` python
        try:
            return self.find_plugin(name, collection_list=collection_list) is not None
        except Exception as ex:
            display.warn('find_plugin(name=%s, collection_list=%s) raised %s' % (name, collection_list, ex))
```



In my dev branches with logging setup, the following works a lot better for me, but less useful with stock ansible:

``` python
    def has_plugin(self, name, collection_list=None):
        ''' Checks if a plugin named name exists '''

        try:
            return self.find_plugin(name, collection_list=collection_list) is not None
        except Exception as ex:
            self.log.exception(ex)
            self.log.debug('find_plugin(name=%s, collection_list=%s) raised %s', name, collection_list, ex)
            pass
```